services/motion/builtin/state/state.go

```diff
@@ -0,0 +1,619 @@
+	}
+	lastPlanWithStatus.StatusHistory = append([]motion.PlanStatus{update.planStatus}, lastPlanWithStatus.StatusHistory...)
+	// write updated last plan back to history
+	execution.history[0] = lastPlanWithStatus
+	// write the execution with the new history to the component execution state copy
+	componentExecutions.executionsByID[update.executionID] = execution
+	// write the component execution state copy back to the state
+	s.componentStateByComponent[update.componentName] = componentExecutions
+
+	return nil
+}
+
+func (s *State) activeExecution(name resource.Name) (stateExecution, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	if cs, exists := s.componentStateByComponent[name]; exists {
+		es := cs.lastExecution()
+
+		if _, exists := motion.TerminalStateSet[es.history[0].StatusHistory[0].State]; exists {
+			return stateExecution{}, ErrNotFound
+		}
+		return es, nil
+	}
+	return stateExecution{}, ErrUnknownResource
+}
+
+// ExecutionReq is the request type for StartExecution.
+type ExecutionReq struct {
+	ComponentName      resource.Name
+	Destination        *geo.Point
+	Heading            float64
+	MovementSensorName resource.Name
+	Obstacles          []*spatialmath.GeoObstacle
+	MotionCfg          *motion.MotionConfiguration
+	Extra              map[string]interface{}
+}
+
+// NewExecutionReq creates an ExecutionReq from a motion.MoveOnGlobeReq.
+func NewExecutionReq(req motion.MoveOnGlobeReq) ExecutionReq {
+	return ExecutionReq{
+		ComponentName:      req.ComponentName,
+		Destination:        req.Destination,
+		Heading:            req.Heading,
+		MovementSensorName: req.MovementSensorName,
+		Obstacles:          req.Obstacles,
+		MotionCfg:          req.MotionCfg,
+		Extra:              req.Extra,
+	}
+}
+
+// execution represents the state of a motion planning execution.
+// it only ever exists in state.StartExecution function & the go routine created
+// its `start()` method.
+type execution struct {
+	// only the id is exposed as the rest are implementation details
+	id            motion.ExecutionID
+	state         *State
+	componentName resource.Name
+	waitGroup     *sync.WaitGroup
+	cancelCtx     context.Context
+	cancelFunc    context.CancelFunc
+	testConfig    *TestConfig
+	logger        golog.Logger
+}
+
+// NewPlan creates a new motion.Plan from an execution & returns an error if one was not able to be created.
+func (e *execution) newPlan() (motion.Plan, error) {
+	// TODO: Generate steps
+	if e.testConfig.newPlanFailReason != nil {
+		return motion.Plan{}, errors.New(*e.testConfig.newPlanFailReason)
+	}
+	return motion.Plan{ID: uuid.New(), ExecutionID: e.id, ComponentName: e.componentName}, nil
+}
+
+// Start starts an execution with a given plan.
+func (e *execution) start(originalPlan motion.Plan) error {
+	if err := e.notifyStateNewExecution(e.toStateExecution(), originalPlan, time.Now()); err != nil {
+		return err
+	}
+
+	e.state.waitGroup.Add(1)
+	e.waitGroup.Add(1)
+	go func() {
```

Comment:
IMO `ManagedGo` should only be used where it's really necessary; it will, on a panic, re-call the `go()`'d to start again from scratch, which can have some really nasty side effects. It's not obvious from the name that that's what that does; executing half of a function, panicking, and then re-executing the first half of that same function.